x/protorev/keeper/rebalance.go

```diff
@@ -0,0 +1,160 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	gammtypes "github.com/osmosis-labs/osmosis/v13/x/gamm/types"
+	"github.com/osmosis-labs/osmosis/v13/x/protorev/types"
+)
+
+// IterateRoutes checks the profitability of every single route that is passed in
+// and returns the optimal route if there is one
+func (k Keeper) IterateRoutes(ctx sdk.Context, routes []gammtypes.SwapAmountInRoutes) (sdk.Coin, sdk.Int, gammtypes.SwapAmountInRoutes) {
+	var optimalRoute gammtypes.SwapAmountInRoutes
+	maxProfitInputCoin := sdk.NewCoin(types.OsmosisDenomination, sdk.ZeroInt())
+	maxProfit := sdk.ZeroInt()
+
+	for _, route := range routes {
+		// Find the max profit for the route using the token out denom of the last pool in the route as the input token denom
+		inputCoin, profit, err := k.FindMaxProfitForRoute(ctx, route, route[2].TokenOutDenom)
+		if err != nil {
+			k.Logger(ctx).Error("Error finding max profit for route: ", err)
+			continue
+		}
+
+		// Filter out routes that don't have any profit
+		if profit.LTE(sdk.ZeroInt()) {
+			continue
+		}
+
+		// If arb doesn't start and end with uosmo, then we convert the profit to uosmo, and compare profits in terms of uosmo
+		if inputCoin.Denom != types.OsmosisDenomination {
```

Comment:
Asking to learn: If the arb profit is really high wouldn't we create another arb position in balanced uosmo pool while converting? Is this expected?

Reply:
@stackman27 So this logic isn't actually swapping the asset to uosmo (which could create an arb opportunity). This logic just called CalcOutAmtGivenIn and sees what the converted profit would be if it was osmo-denominated, and chooses the highest route/tokenIn that gives the most osmo-denominated profits (but still in the original denom)